source/mir/algorithm/iteration.d

```diff
@@ -3891,3 +3892,112 @@ version(mir_test)
     auto rc = matrix.byDim!1.map!filterPositive;
     assert(equal!equal(rc, [ [3, 100], [], [400, 102] ]));
 }
+
+/++
+Implements the homonym function (also known as `accumulate`, $(D
+compress), `inject`, or `foldl`) present in various programming
+languages of functional flavor. The call `fold!(fun)(slice, seed)`
+first assigns `seed` to an internal variable `result`,
+also called the accumulator. Then, for each element `x` in $(D
+slice), `result = fun(result, x)` gets evaluated. Finally, $(D
+result) is returned.
+
+Params:
+    fun = the predicate function to apply to the elements
+
+See_Also:
+    $(HTTP en.wikipedia.org/wiki/Fold_(higher-order_function), Fold (higher-order function))
+
+    $(LREF sum) is similar to `fold!((a, b) => a + b)` that offers
+    precise summing of floating point numbers.
+
+    This is functionally equivalent to $(LREF reduce) with the argument order
+    reversed, and without the need to use $(REF_ALTTEXT `tuple`,tuple,std,typecons)
+    for multiple seeds.
++/
+template fold(alias fun)
+{
+    /++
+    Params:
+        slice = A slice, range, and array.
+        seed = An initial accumulation value (optional).
+    Returns:
+        the accumulated result
+    +/
+    @optmath auto fold(Slice, S)(scope Slice slice, S seed)
+        if (hasLength!Slice || isInputRange!Slice)
+    {
+        static if (hasLength!Slice) {
+            assert(slice.length > 0, "fold: slice must have positive length");
+        } else {
+            assert(!slice.empty, "fold: slice must not be empty");
```

Comment:
> In addition, it should throw an exception in case of an empty input.

This comment was related only to a single argument overload. It is OK to have an empty slice with two argument overload.